File: AlgoExpert_Iterative In-order Traversal.cpp
#include <bits/stdc++.h>
using namespace std;

class BinaryTree {
public:
  int value;
  BinaryTree *left;
  BinaryTree *right;
  BinaryTree *parent;

  BinaryTree(int value, BinaryTree *parent = nullptr);
  void insert(vector<int> values, int i = 0);
};

void iterativeInOrderTraversal(BinaryTree *tree, void (*callback)(BinaryTree *tree)) {
    BinaryTree * previeousNode = nullptr;
    BinaryTree * currentNode = tree;
    BinaryTree * nextNode = nullptr;

    while (currentNode != nullptr)
    {
        if(previeousNode == nullptr || previeousNode == currentNode->parent) {
            if(currentNode->left != nullptr) {
                nextNode = currentNode->left;
            } else {
                (*callback)(currentNode);
                nextNode = (currentNode->right != nullptr ? currentNode->right : currentNode->parent);
            }
        }

        else if(previeousNode == currentNode->left) {
            (*callback)(currentNode);
            nextNode = (currentNode->right != nullptr ? currentNode->right : currentNode->parent);
        }

        else {
            nextNode = currentNode->parent;
        }

        previeousNode = currentNode;
        currentNode = nextNode;
    }
    
    return ;
}

File: AlgoExpert_Iterative_In_order_Traversal.java
import java.util.function.Function;


public class AlgoExpert_Iterative_In_order_Traversal {
    public static void iterativeInOrderTraversal(BinaryTree tree, Function<BinaryTree, Void> callback) {
        BinaryTree parent = null;
        BinaryTree current = tree;
        BinaryTree next = null;
        while(current != null) {
            if(parent == null || parent == current.parent) {
                if(current.left != null) {
                    next = current.left;
                } else {
                    callback.apply(current);
                    next = (current.right == null ? current.parent : current.right);
                }
            } else if(parent == current.left) {
                callback.apply(current);
                next = (current.right == null ? current.parent : current.right);
            } else {
                next = current.parent;
            }
            parent = current;
            current = next;
        }

    }
  
    static class BinaryTree {
      public int value;
      public BinaryTree left;
      public BinaryTree right;
      public BinaryTree parent;
  
      public BinaryTree(int value) {
        this.value = value;
      }
  
      public BinaryTree(int value, BinaryTree parent) {
        this.value = value;
        this.parent = parent;
      }
    }
}
